src/LCM2ROSControl.cpp

```diff
@@ -204,43 +272,79 @@ namespace valkyrie_translator
+          lcm_torque_msg.joint_position[effortJointIndex] = command_effort;
+
+          effortJointIndex++;
+      }
+
+      // Iterate over all position-controlled joints
+      size_t positionJointIndex = effortJointIndex;
+      for (auto iter = positionJointHandles.begin(); iter != positionJointHandles.end(); iter++) {
+          double q = iter->second.getPosition();
+          double qd = iter->second.getVelocity();
+
+          joint_command& command = latest_commands[iter->first];
+          double position_to_go = command.position;
+
+          // TODO: check that desired q is within limits
+          if (fabs(position_to_go) < M_PI) { // TODO: check joint limit!
+            iter->second.setCommand(position_to_go);
+          } else{
+            ROS_INFO("Dangerous latest_commands[%s]: %f\n", iter->first.c_str(), position_to_go);
+            // iter->second.setCommand(0.0); // Don't do anything!
+          }
+          // TODO: we can't directly iterate like this, better match differently!!
+
+          lcm_pose_msg.joint_name[positionJointIndex] = iter->first;
+          lcm_pose_msg.joint_position[positionJointIndex] = q;
+          lcm_pose_msg.joint_velocity[positionJointIndex] = qd;
+
+          lcm_state_msg.joint_name[positionJointIndex] = iter->first;
+          lcm_state_msg.joint_position[positionJointIndex] = q;
+          lcm_state_msg.joint_velocity[positionJointIndex] = qd;
 
           // republish to guarantee sync
-          lcm_commanded_msg.joint_name[i] = iter->first;
-          lcm_commanded_msg.joint_position[i] = command.position;
-          lcm_commanded_msg.joint_velocity[i] = command.velocity;
-          lcm_commanded_msg.joint_effort[i] = command.effort;
+          lcm_commanded_msg.joint_name[positionJointIndex] = iter->first;
+          lcm_commanded_msg.joint_position[positionJointIndex] = command.position;
+          lcm_commanded_msg.joint_velocity[positionJointIndex] = command.velocity;
+          lcm_commanded_msg.joint_effort[positionJointIndex] = command.effort;
 
-          lcm_torque_msg.joint_name[i] = iter->first;
-          lcm_torque_msg.joint_position[i] = command_effort;
+          positionJointIndex++;
+      }
 
-          i++;
-      }   
       lcm_->publish("VAL_CORE_ROBOT_STATE", &lcm_pose_msg);
       lcm_->publish("VAL_COMMAND_FEEDBACK", &lcm_commanded_msg);
       lcm_->publish("VAL_COMMAND_FEEDBACK_TORQUE", &lcm_torque_msg);
       lcm_->publish("EST_ROBOT_STATE", &lcm_state_msg);
 
       // push out the measurements for all imus we see advertised
+      int i;
       for (auto iter = imuSensorHandles.begin(); iter != imuSensorHandles.end(); iter ++){
         bot_core::ins_t lcm_imu_msg;
         //lcm_imu_msg.utime = utime;
         std::ostringstream imuchannel;
         imuchannel << "VAL_IMU_" << iter->first;
         lcm_imu_msg.utime = utime;
-        for (i=0; i<3; i++){
+        for (int i=0; i<3; i++){
```

Comment:
Good catch - probably a bug, need to check this (in sim we don't get IMU data and on the robot we haven't checked it so far).